### PR TITLE
Optimisations for quoting / unquoting

### PR DESCRIPTION
**Pull request by Sanne**
- https://hibernate.atlassian.net/browse/HHH-13512

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/test/util/StringHelperTest.java [hunk 1]
```diff
@@ -54,4 +60,33 @@ public void testFindIdentifierWord() {
...
+	private static H2Dialect DIALECT = new H2Dialect();
+	@Test
+	public void testArrayUnquoting() {
+		assertNull( StringHelper.unquote( (String[]) null, DIALECT ) );
+		//This to verify that the string array isn't being copied unnecessarily:
+		unchanged( new String [0] );
+		unchanged( new String[] { "a" } );
+		unchanged( new String[] { "a", "b" } );
+		helperEquals( new String[] { "a", "b", "c" }, new String[] { "a", "b", "`c`" } );
+		helperEquals( new String[] { "a", "b", "c" }, new String[] { "a", "\"b\"", "c" } );
```
Comment (dreab8) [inline hibernate-core/src/test/java/org/hibernate/test/util/StringHelperTest.java]: i would add `helperEquals( new String[] { "\"a", "b\"", "`c", "d`" }, new String[] { "\"a", "b\"", "`c", "d`" } );`

File: hibernate-core/src/main/java/org/hibernate/internal/util/StringHelper.java [hunk 2]
```diff
@@ -739,15 +742,32 @@ public static String unquote(String name, Dialect dialect) {
...
+			//In this case all strings are already unquoted, so return the same array as the input:
+			//this is a good optimisation to skip an array copy as typically either all names are consistently quoted, or none are;
+			//yet for safety we need to deal with mixed scenarios as well.
+			return names;
+		}
+		else {
+			String[] unquoted = new String[length];
+			System.arraycopy( names, 0, unquoted, 0, failedIndex );
+			for ( int i = failedIndex; i < length; i++ ) {
+				unquoted[i] = unquote( names[i], dialect );
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/internal/util/StringHelper.java]: in such a case isQuoted(..) is executed twice, not sure about the impact.

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/internal/util/StringHelper.java]: sorry my mistake it is called only 1 time more, forget the comment

**Comment by Sanne** [comment]
merged

**Pull request closed without merging**